### DISCUSSION TITLE
Broken export

### DISCUSSION THREAD

**Discussion by nitriques** [solspace/craft-freeform#870] [Questions & Answers] [answered]
**Describe the bug or issue you're experiencing**

I get a fatal error when going to `/craft/freeform/export/profiles`

**Steps to reproduce**

1. Go to export in freeform
2. I get a 500. Here's the stack strace:

```
TypeError: Solspace\Freeform\Elements\Submission::getFieldColumnName(): Argument #1 ($field) must be of type Solspace\Freeform\Library\Composer\Components\FieldInterface, null given, called in /dev/vendor/solspace/craft-freeform/packages/plugin/src/Models/Pro/ExportProfileModel.php on line 338 and defined in /dev/vendor/solspace/craft-freeform/packages/plugin/src/Elements/Submission.php:257
Stack trace:
#0 /dev/vendor/solspace/craft-freeform/packages/plugin/src/Models/Pro/ExportProfileModel.php(338): Solspace\Freeform\Elements\Submission::getFieldColumnName(NULL)
#1 /dev/vendor/solspace/craft-freeform/packages/plugin/src/Models/Pro/ExportProfileModel.php(76): Solspace\Freeform\Models\Pro\ExportProfileModel->buildCommand()
#2 /dev/vendor/yiisoft/yii2/base/Component.php(139): Solspace\Freeform\Models\Pro\ExportProfileModel->getSubmissionCount()
#3 /dev/vendor/craftcms/cms/src/helpers/Template.php(118): yii\base\Component->__get('submissionCount')
#4 /dev/storage/runtime/compiled_templates/fa/fa28774b286bac65a7ba5707e81a280d.php(141): craft\helpers\Template::attribute(Object(craft\web\twig\Environment), Object(Twig\Source), Object(Solspace\Freeform\Models\Pro\ExportProfileModel), 'submissionCount', Array)
#5 /dev/vendor/twig/twig/src/Template.php(394): __TwigTemplate_922d17776da8ede98d617adf0b0cbb0f->doDisplay(Array, Array)
#6 /dev/vendor/twig/twig/src/Template.php(367): Twig\Template->displayWithErrorHandling(Array, Array)
#7 /dev/vendor/twig/twig/src/Template.php(379): Twig\Template->display(Array)
#8 /dev/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\Template->render(Array, Array)
#9 /dev/vendor/twig/twig/src/Environment.php(277): Twig\TemplateWrapper->render(Array)
#10 /dev/vendor/craftcms/cms/src/web/View.php(474): Twig\Environment->render('freeform/export...', Array)
#11 /dev/vendor/craftcms/cms/src/web/View.php(527): craft\web\View->renderTemplate('freeform/export...', Array)
#12 /dev/vendor/craftcms/cms/src/web/TemplateResponseFormatter.php(56): craft\web\View->renderPageTemplate('freeform/export...', Array, 'cp')
#13 /dev/vendor/yiisoft/yii2/web/Response.php(1100): craft\web\TemplateResponseFormatter->format(Object(craft\web\Response))
#14 /dev/vendor/craftcms/cms/src/web/Response.php(337): yii\web\Response->prepare()
#15 /dev/vendor/yiisoft/yii2/web/Response.php(340): craft\web\Response->prepare()
#16 /dev/vendor/yiisoft/yii2/base/Application.php(390): yii\web\Response->send()
#17 /dev/web/index.php(21): yii\base\Application->run()
#18 {main}
```

**Expected behavior**

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 4.5.5
- Freeform Version: 4.1.8
- Freeform Edition: Pro
- Fresh Install or Upgrade: Upgrade

**Additional context**

I do not have the error with other DBs, looks like a data corruption following a delete or rename of a field in a form. But either way, it should not fail that hard.

I tried to run craft's gc to no avail.

**Comment by kjmartens** [comment]
Sorry for the trouble you're experiencing, @nitriques. Can you check the `freeform_fields` database table and let me know if any are missing a value in the `type` column? If everything appears to be okay, can you send us a copy of the Freeform database tables (at least this one) for review in a [private support ticket](https://docs.solspace.com/craft/freeform/v4/support/#create-a-support-ticket)? 🙂

**Comment by nitriques** [comment]
Hi @kjmartens Thanks for the reply. I might add that this is a _big_ website, which gets around 50,000 submissions per month. Just managing the MySQL database is a pain.

```
select * from craft_freeform_fields where type is null
--> 0 result

select * from craft_freeform_fields where type = ''
--> 0 result

```

There are 103 fields, and they all look "ok"

**Comment by kjmartens** [comment]
Ok @nitriques. I wonder if any of them have an incorrect value somehow. Can you send us a copy of the `freeform_fields` database table for review? 🙂

**Comment by nitriques** [comment]
[query-1.sql.txt](https://github.com/solspace/craft-freeform/files/12730418/query-1.sql.txt)

There you go !

**Comment by kjmartens** [comment]
Thanks @nitriques. Indeed, nothing appears to be out of sorts there.

Just to confirm, this happens as soon as you visit the Export Profile page? Or does it happen when you attempt to download an export file for a form?

How many export profiles do you have set up? If it's just one or a couple, could you copy and paste the JSON for the `layoutJson` column of the form(s) that have export profiles set up in the `freeform_forms` database table? 🙂 

I would also be curious what the structure/columns of any affected form's `freeform_submissions_form_name_123` tables are. Perhaps there's an orphaned field column in there.

I apologize for this getting a bit messy, but I have also never seen this issue come up before, unfortunately.

**Comment by nitriques** [comment]
Hi @kjmartens 

> Just to confirm, this happens as soon as you visit the Export Profile page? Or does it happen when you attempt to download an export file for a form?

As soon as I visit the Export Profile page.

I'll come back ASAP with all the details you requested.


> I apologize for this getting a bit messy, 

That's sooooooo not your fault ;) Thanks for you help, I appreciate it.

**Comment by nitriques** [comment]
@kjmartens 

1. The `freeform_submissions_form_name_123 (craft_freeform_submissions_form_name_123)` table does not exists!!
2. In which table should the `layoutJson` column reside ? I have 129 forms...

**Comment by kjmartens** [comment]
Hi @nitriques,

The `freeform_submissions_form_name_123` table name was an example only. I was suggesting that if, for example, you had a form named "Contact Form" (with an ID of 123), can you review the matching submissions table for it, e.g. `freeform_submissions_contact_form_123`. But your table names will differ of course. 🙂 

As for where to find `layoutJson`, it's located in the `freeform_forms` database table. A row for each form will exist there. 

In either of these cases, being that you have so many forms, I was wondering if you had any idea how many Export Profiles you had set up, as it could help narrow things down. 🙂

**Comment by nitriques** [comment]
Hi @kjmartens 

> But your table names will differ of course. 🙂

Ah sorry! Well yes, I have many of those tables. I just do not know which form is causing the error.

> I was wondering if you had any idea how many Export Profiles

I do not know, I can not see them ! I would say less then 10...
Are there in the DB ?

> As for where to find layoutJson, it's located in the freeform_forms database table. A row for each form will exist there.

Yeah ok I was able to find it. But they all look alike, and they all look like valid json.

**Comment by kjmartens** [comment]
> > I was wondering if you had any idea how many Export Profiles
> 
> I do not know, I can not see them ! I would say less then 10... Are there in the DB ?

You would need to review (or export and provide me with) the `freeform_export_profiles` table for this (`fields` JSON).

> > As for where to find layoutJson, it's located in the freeform_forms database table. A row for each form will exist there.
> 
> Yeah ok I was able to find it. But they all look alike, and they all look like valid json.

They very likely are valid JSON. However, my thinking is that you might have a stray/orphaned field in there that doesn't actually exist, but is causing this issue. It'll be a needle in a haystack to compare the tables and see if/where this might have broken down.

If you recall some recent field names you've deleted, that could help in the search. 🙂

**Comment by nitriques** [comment]
Hi @kjmartens 

Ok thanks. Sadly, I do not work on the content management (I am a dev) and the team does not know.

Here's the profile export: [craft_freeform_export_profiles.sql.txt](https://github.com/solspace/craft-freeform/files/12803243/craft_freeform_export_profiles.sql.txt)

> They very likely are valid JSON. However, my thinking is that you might have a stray/orphaned field in there that doesn't actually exist, but is causing this issue. 

Could this be part of the cli ? I mean, I won't test them manually, I would right code to find it ?

Also, I would love for the page to NOT crash if there's data corruption.

Let me know what solution you like, I'll find the budget to finance it.

**Comment by kjmartens** [comment]
Hi @nitriques,

I spent a little bit of time combing through your database table. What I was able to find is that both Field 89 and Field 91 appear to have no label for them. In the `freeform_fields` table, they are set to `OtherReason` and `Autre Raison`, respectively. This also makes sense since it appears they're both the same purpose but with English and French.

I imagine what might have happened is that someone removed the labels for them in the form builder so they don't appear on the rendered form.

Can you try adding labels back for them inside the form builder and see if the issue goes away? The affected forms have the ID's `111` and `112`. 🙂

**Reply by nitriques** [reply]
Hi @kjmartens 

Great, I'll try this asap. So the fields are not deleted, they have an empty label ? Why can we have empty labels if it makes the export crash ?

**Reply by nitriques** [reply]
Hi @kjmartens 

I went ahead and checked those 2 fields and their label is still there.
Then, I went to check the forms using those fields, and indeed, their labels where removed.

Now, since this is allowed in the UI, I would expect the export feature to support it no ?

It should also use the label from the field if the label in the form is empty no ?

Also, the after adding the labels back, the export feature is still broken.

I do not know why this has been move to a discussion: it's a bug, and a nasty one.

**Reply by nitriques** [reply]
Hi @kjmartens sorry for the ping, the our exports are still not working...

**Comment by nitriques** [comment, marked as answer]
Hi @kjmartens 

I was able to monkey patch it, in order to get the profile page loading again (but with a broken submission count), like so:

<img width="782" alt="image" src="https://github.com/solspace/craft-freeform/assets/771169/53f0eb1e-0cbc-4c1d-90cc-53c5d8538bce">

The bug is in the submission count code. For some reason, it can't find the field it's looking for (surely it was deleted/removed?). In the export profile, I was able to find the field id `114`, which is not longer part of the form. The field _still_ exists, just not used in the from linked to that export profile.

The proper fix would be to make sure that filters are _cleaned up_ upon each form save. Filter created on non-linked field should not make the whole system crash: at the very least, the should be ignored like I did (but this is an ugly fix by itself).

With the fix, I was also able to go into the profile details page and this is what I saw (with the label being empty)

<img width="447" alt="image" src="https://github.com/solspace/craft-freeform/assets/771169/c744bdb4-6829-428a-99c7-622b0d0dc155">

This looks like the _original_ problem: it should use the field "main/default" label when its empty in the form. But this is not what the fatal error is about. IMHO, both should be properly addressed.

---

At least, resaving the broken export profiles made the page work again, but I fear like it still can break. Somehow, the profile detail page ignored the non-linked field ? We are still not sure about how the export notification will react to all of this. More to follow.

**Reply by kjmartens** [reply]
I apologize for the delay on this. And thank you for the additional information.

I have made a note about this and a developer will investigate the issue and make any necessary changes. 🙂

**Reply by nitriques** [reply]
Thanks @kjmartens

**Reply by nitriques** [reply]
@kjmartens Any news on this issue ? We also ran on other problems where the exports would need 1.5Gb of memory to render. Is there a way to render the export via Craft's queue ?

**Reply by kjmartens** [reply]
We addressed the issue of the field not being removed from the Export Profile when removed from the builder. That is available in Freeform 4.1.12+. 🙂 

https://docs.solspace.com/craft/freeform/v4/setup/changelog/#_4-1-12-2023-11-09

As for the memory issue you're running into, I'm sorry about that. I will make a note about that for future consideration. 🙂

**Reply by nitriques** [reply]
Hi @kjmartens 

Wow thanks great, thanks ! Must have missed the release.

This big client of ours is running at a really big scale. 1M+ page views per days, sometimes as much as 4000 entries per days in the forms.

They might also contact you guys directly, to talk about financing some features they need (this would not be their first time, but its a new web team)

Thanks again for your amazing support, we really appreciate it.
